### ISSUE TITLE
Revisit: full nested split for ApproximateDistribution accessors (Option A)

### ISSUE THREAD

**Issue by jhuggins** [TARPS-group/prob-pipe#164]
## Context

PR #155 (PR-B) shipped Option B for nested-template `ApproximateDistribution`: when `record_template` has nested `RecordTemplate` fields, `__init__` slices the flat chain by **top-level** template field. Nested fields are stored as a flat `(n, nested_flat_size)` array; the nested structure is recoverable via `record_template[field]` and `draws()`.

That fixes the user-visible asymmetry (`mean(post)` keys now match `post.fields`), but trades off the resolution at the moment / event-shape layer:

```python
template = RecordTemplate(
    params=RecordTemplate(a=(), b=()),
    scale=(),
)
# ...
post.event_shapes               # {'params': (2,), 'scale': ()}     — flat per top-level
mean(post)['params']            # array shape (2,)                  — flat per top-level
```

Users who want to inspect the nested structure use `record_template['params']` or `draws()['params']['a']`. That's a deliberate consistency choice — `event_shapes` reports one shape per `field`, and `field` is `'params'`, not `'params/a'`.

## Question for this issue

Should we extend the constructor to do a **full recursive split**, materialising nested `NumericRecord` leaves so the per-leaf structure is exposed at every accessor?

**Option A behavior** (the alternative we deferred):

```python
# _record_data['params'] would be NumericRecord(a=array(40,), b=array(40,))
post.event_shapes  # {'params': {'a': (), 'b': ()}, 'scale': ()}  — recursive
mean(post)         # NumericRecord(params=NumericRecord(a=scalar, b=scalar), scale=scalar)
mean(post)['params']['a']  # scalar mean of a
```

## Why deferred

Cost was significant for a review-fix PR:

- ~10 accessors in `RecordEmpiricalDistribution` (`event_shapes`, `event_shape`, `dtypes`, `dim`, `_sample`, `_mean`, `_variance`, `_cov`, `flat_samples`, `_expectation`) currently do `jnp.asarray(self._record_data[f])`. That fails on a nested `NumericRecord` field via the single-field shim's `_single_numeric_leaf` raise. Each accessor needs to dispatch on whether `f` is a flat-leaf or nested-NumericRecord.
- `_validate_record_samples` (`probpipe/core/_empirical.py`) needs to recurse — currently it does `jnp.asarray(record_data[f]).shape[0]` per field, expecting flat arrays.
- Return type changes for `event_shapes`, `dtypes`, `_mean`, etc. from `dict[str, tuple]` / `dict[str, Array]` to a recursive form. Existing tests assert on the flat-dict types and would need updates.

Estimated effort: 50-80 lines across `_empirical.py` plus test updates, ~3-4 hours of careful work.

## Triggers for revisiting

Reasons to do Option A later:

1. **A real consumer surfaces** that wants per-leaf moments without going through `draws()`. Most likely candidate: an arviz-style summary helper that reports parameter-by-parameter posterior stats.
2. **A user asks** for `mean(post)['params']['a']` to work directly (without `draws()`).
3. **`flat_samples` for nested templates becomes ambiguous** — currently the `(n, total_dim)` flat view is unambiguous because we collapse nesting at the data layer. Option A might want to expose a nested moment view alongside the flat one.

Reasons to leave it:

1. **`draws()` already gives users the nested structure** — `post.draws()['params']['a']` works today.
2. **`record_template` exposes the nested template structure** for introspection.
3. **The flat per-top-level view of moments is what most MCMC posterior-inspection actually wants** — users typically summarise per top-level parameter group, not per leaf.
4. **The current fix is consistent and well-defined** — every accessor returns one entry per `post.fields[i]`.

## Tracking

- Pin behavior: `tests/test_inference.py::TestApproximateDistributionValuesTemplate::test_nested_template_accessors_match_top_level_fields`.
- Constructor: `probpipe/inference/_approximate_distribution.py:ApproximateDistribution.__init__`.
- The fix shipped in PR #155 reverts cleanly to Option A — the change is contained inside `__init__`.

## Acceptance criteria for closing this issue

Either:

- Implement Option A: every accessor on `ApproximateDistribution` (and by extension `RecordEmpiricalDistribution`) returns a recursive structure matching the template's nested fields. Update tests. Decide on the new return types for `event_shapes` / `dtypes` / `_mean` / etc.

OR:

- Confirm Option B is sufficient long-term: no real consumer has surfaced asking for per-leaf moments, the flat per-top-level view is the desired summary. Close as won't-fix.

**Comment by jhuggins** [comment]
Closing as superseded: `ApproximateDistribution` dissolves under the design reference — an inference result is an ordinary member of whichever family realizes it, with approximation recorded in `provenance` (`design/06-distribution-catalog.md`, inference-produced distributions; tracked in #342). The nested-accessor semantics this issue asks for arrive anyway through the operation contracts: `mean` / `variance` return a `Record` matching the event template (`design/05-operations.md`).
